### PR TITLE
[Webpack 5]: Storybook Fixes

### PR DESCRIPTION
**Pull request by fallaciousreasoning**
Resolves https://github.com/brave/brave-browser/issues/34705

Changes:
1. Don't use Typescript for `chrome-resources-mocks` - this doesn't play nice with the `NormalModuleReplacementPlugin`
2. Rename `cr.m.js.ts` to `cr.js` - we don't import `cr.m.js` anywhere, so the mock was never being used
3. Decided at the start of the build what output folder will be used, rather than searching all build output folders for every file.
4. Use `path.join` instead of `require.resolve` - I think this was the cause of the wrong build output folder being used

## Submitter Checklist:

- [ ] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/reviews/issues/new/choose) them
- [ ] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [ ] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [ ] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [ ] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run lint`, `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

File: .storybook/webpack.config.js [hunk 1]
```diff
@@ -20,28 +20,28 @@ function getBuildOuptutPathList(buildOutputRelativePath) {
...
 // Mock ROOT_GEN_DIR as 'gen' - it will be replaced with an array from
 // getBuildOutputPathList.
-process.env.ROOT_GEN_DIR = 'gen'
+const potentialGenFolders = getBuildOuptutPathList('gen')
+process.env.ROOT_GEN_DIR = potentialGenFolders
+  .filter(a => fs.existsSync(a))
+  .sort((a, b) => fs.statSync(a).mtime - fs.statSync(b).mtime)[0]
+
+  // Fallback to the first folder in the list, so we get sensible errors.
+  ?? potentialGenFolders[0]
```
Comment (petemill) [inline .storybook/webpack.config.js:29]: If none of them exist, shouldn't we just throw an error here?

Comment (fallaciousreasoning) [reply .storybook/webpack.config.js:29]: That might be sensible - I'll update.

File: .storybook/webpack.config.js [hunk 2]
```diff
@@ -20,28 +20,28 @@ function getBuildOuptutPathList(buildOutputRelativePath) {
...
+
 const basePathMap = require('../components/webpack/path-map')
 
 // Override the path map we use in the browser with some additional mock
 // directories, so that we can replace things in Storybook.
 const pathMap = {
   ...basePathMap,
-  gen: [
```
Comment (fallaciousreasoning) [inline .storybook/webpack.config.js:30]: Gen mock was not being used.

File: .storybook/webpack.config.js [hunk 3]
```diff
@@ -20,28 +20,28 @@ function getBuildOuptutPathList(buildOutputRelativePath) {
...
-process.env.ROOT_GEN_DIR = 'gen'
+const potentialGenFolders = getBuildOuptutPathList('gen')
+process.env.ROOT_GEN_DIR = potentialGenFolders
+  .filter(a => fs.existsSync(a))
+  .sort((a, b) => fs.statSync(a).mtime - fs.statSync(b).mtime)[0]
+
+  // Fallback to the first folder in the list, so we get sensible errors.
+  ?? potentialGenFolders[0]
+
 const basePathMap = require('../components/webpack/path-map')
```
Comment (petemill) [inline .storybook/webpack.config.js:28]: probably a less hacky way than making sure the env variable is defined before the first `require` of `path-map.js` is to have path-map return a function that takes a prefix and have both webpacks pass in different variables. But at least it works for now.

File: .storybook/webpack.config.js (hunk 1 above)
Comment (petemill) [reply .storybook/webpack.config.js]: I think that would be sensible so people know they can't generate Storybook without a build first.

**Review by petemill — APPROVED** [review]

File: components/webpack/path-map.js [hunk 4]
```diff
@@ -5,32 +5,38 @@
...
+  return {
+    // Find files in the current build configurations /gen directory
+    'gen': genPath,
+    // Generated resources at this path are available at chrome://resources and
+    // whilst webpack will still bundle, we keep the alias to the served path
+    // to minimize knowledge of specific gen/ paths and easily allow us to not
+    // bundle
+    // them in the future for certain build configuration, just like chromium.
+    'chrome://resources/brave/fonts': path.join(
+      genPath,
```
[the remark this answers is not in the document]
Comment (petemill) [reply components/webpack/path-map.js:22]: This is not changing anything, and this is a build script. Just changing from env var to regular var (provided by env var).

File: components/webpack/path-map.js [hunk 5]
```diff
@@ -5,32 +5,38 @@
...
+      'brave/ui/webui/resources/fonts'),
+    'chrome://resources/brave': path.join(
+      genPath,
```
[the remark this answers is not in the document]
Comment (petemill) [reply components/webpack/path-map.js:25]: same

File: components/webpack/path-map.js [hunk 6]
```diff
@@ -5,32 +5,38 @@
...
+      'brave/ui/webui/resources/tsc'),
+    'chrome://resources': path.join(
+      genPath, 'ui/webui/resources/tsc'),
```
[the remark this answers is not in the document]
Comment (petemill) [reply components/webpack/path-map.js:28]: same

File: .storybook/webpack.config.js [hunk 7]
```diff
@@ -18,30 +23,27 @@ function getBuildOuptutPathList(buildOutputRelativePath) {
...
-process.env.ROOT_GEN_DIR = 'gen'
-const basePathMap = require('../components/webpack/path-map')
+const genFolder = getBuildOuptutPathList('gen')
+  .filter(a => fs.existsSync(a))
+  .sort((a, b) => fs.statSync(b).mtime - fs.statSync(a).mtime)[0]
+if (!genFolder) {
+  throw new Error("Failed to find build output folder!")
+}
+
+const basePathMap = require('../components/webpack/path-map')(genFolder)
```
Comment (petemill) [inline .storybook/webpack.config.js:33]: well, now we can put that `require` at the top of the file

**Pull request merged**